### PR TITLE
mds: do not take the ino which has been used

### PR DESCRIPTION
**Pull request by lxbsz**
When replaying the journals, if the inodetable or the sessionmap
versions do not match, the CInode will be added to the inode_map,
but the ino may still be in the inodetable or sessions' prealloc
inos list.

So when allocating new CInode we should skip them.

Fixes: https://tracker.ceph.com/issues/52280
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mds/MDCache.h [hunk 1]
```diff
@@ -1215,6 +1222,8 @@ class MDCache {
   StrayManager stray_manager;
 
  private:
+  std::set<inodeno_t> replay_taked_inos; // the inos have been taken when replaying
```
Comment (jtlayton) [inline src/mds/MDCache.h:1225]: nit: "taked" is not a word in English. Can you change this to "taken" ?

**Review by jtlayton — CHANGES_REQUESTED** [review]
Aside from the naming nit, this looks fine to me.

File: src/mds/MDCache.h (hunk 1 above)
Comment (lxbsz) [reply src/mds/MDCache.h]: Done, thanks :-)

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
Xiubo is looking into adding a test for this.

**Comment by lxbsz** [comment]
> Xiubo is looking into adding a test for this.

Finished coding and I will run the teuthology test today.

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Comment by yiyiarrow** [comment]
In API "CDentry *CDir::_load_dentry"，the CInode will also be added to the inode_map, but the ino may still be in the inodetable or sessions' prealloc inos list. Is it the same problem?

**Comment by lxbsz** [comment]
> In API "CDentry *CDir::_load_dentry"，the CInode will also be added to the inode_map, but the ino may still be in the inodetable or sessions' prealloc inos list. Is it the similar situation?

BTW, did you hit any issue in this code ? It is, but I didn't see how will this happen. Do you have logs ?

**Comment by lxbsz** [comment]
Yeah, from logs it's the same issue I have hit. I will try to fix this case too later.

**Comment by lxbsz** [comment]
> > Yeah, from logs it's the same issue I have hit. I will try to fix this case too later.
> 
> Thanks a lot.

I have fixed it in this PR.

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Comment by yiyiarrow** [comment]
> > > Yeah, from logs it's the same issue I have hit. I will try to fix this case too later.
> > 
> > 
> > Thanks a lot.
> 
> I have fixed it in this PR.

Thanks again.

**Comment by vshankar** [comment]
mostly looks good!

**Review by vshankar — COMMENTED** [review]
@lxbsz Please check these test failures

- https://pulpito.ceph.com/vshankar-2022-01-06_13:18:41-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6598132/
- https://pulpito.ceph.com/vshankar-2022-01-06_13:18:41-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6597968/

**Comment by lxbsz** [comment]
> @lxbsz Please check these test failures
> 
> * https://pulpito.ceph.com/vshankar-2022-01-06_13:18:41-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6598132/
> * https://pulpito.ceph.com/vshankar-2022-01-06_13:18:41-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6597968/

It seems the `mds_kill_skip_replaying_inotable` doesn't successfully set.

**Comment by lxbsz** [comment]
> > @lxbsz Please check these test failures
> > 
> > * https://pulpito.ceph.com/vshankar-2022-01-06_13:18:41-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6598132/
> > * https://pulpito.ceph.com/vshankar-2022-01-06_13:18:41-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6597968/
> 
> It seems the `mds_kill_skip_replaying_inotable` doesn't successfully set.

The mds was restart too fast that the mkdir journal log didn't flushed to disk and lost.

**Comment by vshankar** [comment]
@lxbsz Also causes this test to fail - https://pulpito.ceph.com/vshankar-2022-01-13_04:42:40-fs-wip-vshankar-testing-20220106-145819-testing-default-smithi/6612006/

```
ssertion: /build/ceph-17.0.0-9959-g2747bac1/src/mds/Server.cc: 6336: FAILED ceph_assert(!g_conf()->mds_kill_skip_replaying_inotable)
ceph version 17.0.0-9959-g2747bac1 (2747bac149996f541f92338c26e9f82ec1e3aa8c) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14f) [0x7f6ceed367cc]
 2: /usr/lib/ceph/libceph-common.so.2(+0x27b9de) [0x7f6ceed369de]
 3: (C_MDS_mknod_finish::finish(int)+0x28e) [0x5633e675222e]
 4: (MDSContext::complete(int)+0x61) [0x5633e69bcaa1]
 5: (MDSIOContextBase::complete(int)+0x4ec) [0x5633e69bd1bc]
 6: (MDSLogContextBase::complete(int)+0x45) [0x5633e69bd515]
 7: (Finisher::finisher_thread_entry()+0x19d) [0x7f6ceede69bd]
 8: /lib/x86_64-linux-gnu/libpthread.so.0(+0x9609) [0x7f6ceea66609]
 9: clone()

*** Caught signal (Aborted) **
```

**Comment by vshankar** [comment]
jenkins test api

**Comment by vshankar** [comment]
jenkins test api

**Comment by lxbsz** [comment]
Rebased it.

**Comment by lxbsz** [comment]
jenkins retest this please

**Comment by lxbsz** [comment]
> @lxbsz could you rebase and push an update?

Done and changed the target branch to `ceph:main`.

**Comment by lxbsz** [comment]
jenkins retest this please

**Comment by lxbsz** [comment]
Reopen it and it's ready to review.

**Comment by vshankar** [comment]
jenkins test api

**Comment by vshankar** [comment]
> When replaying the journals, if the inodetable or the sessionmap
> versions do not match, the CInode will be added to the inode_map,
> but the ino may still be in the inodetable or sessions' prealloc
> inos list.

@lxbsz I have lost a bit of context on this issue. If my memory serves me right - when the MDS replays the journal logs the inodes would not be cleaned up from the inodetable, so a client request could reuse the used inode# which results in the crash, right? And how is this exactly related to mismatching inodetable and/or sessionap versions? There is another issue related to sessionmap version mismatch on replay - https://github.com/ceph/ceph/pull/49970. Are both of these related (somehow)?

**Comment by lxbsz** [comment]
> > When replaying the journals, if the inodetable or the sessionmap
> > versions do not match, the CInode will be added to the inode_map,
> > but the ino may still be in the inodetable or sessions' prealloc
> > inos list.
> 
> @lxbsz I have lost a bit of context on this issue. If my memory serves me right - when the MDS replays the journal logs the inodes would not be cleaned up from the inodetable, so a client request could reuse the used inode# which results in the crash, right? 

@vshankar If the mds couldn't replay the journal successfully when the inodetable/sessionmap version mismatched it will skip applying the prealloc `ino#` list, which will remove the `ino#` from this list. But the inode has already been added to the `inode_map` in earlier time when relaying the journal. That means next time when we get a new `ino#` from the prealloc `ino#` list and trying to add it to the `inode_map`, it will conflict.

> And how is this exactly related to mismatching inodetable and/or sessionap versions?

See my above comments.

> There is another issue related to sessionmap version mismatch on replay - #49970. Are both of these related (somehow)?

They are a bit related but different issues.

**Comment by vshankar** [comment]
> > > When replaying the journals, if the inodetable or the sessionmap
> > > versions do not match, the CInode will be added to the inode_map,
> > > but the ino may still be in the inodetable or sessions' prealloc
> > > inos list.
> > 
> > 
> > @lxbsz I have lost a bit of context on this issue. If my memory serves me right - when the MDS replays the journal logs the inodes would not be cleaned up from the inodetable, so a client request could reuse the used inode# which results in the crash, right?
> 
> @vshankar If the mds couldn't replay the journal successfully when the inodetable/sessionmap version mismatched it will skip applying the prealloc `ino#` list, which will remove the `ino#` from this list. But the inode has already been added to the `inode_map` in earlier time when relaying the journal. That means next time when we get a new `ino#` from the prealloc `ino#` list and trying to add it to the `inode_map`, it will conflict.

Sure. But the mismatching versions itself is related to the ceph-mds process restarting without being able to flush mdlog (hitting an assert, etc..)?

**Comment by lxbsz** [comment]
[...]
> > @vshankar If the mds couldn't replay the journal successfully when the inodetable/sessionmap version mismatched it will skip applying the prealloc `ino#` list, which will remove the `ino#` from this list. But the inode has already been added to the `inode_map` in earlier time when relaying the journal. That means next time when we get a new `ino#` from the prealloc `ino#` list and trying to add it to the `inode_map`, it will conflict.
> 
> Sure. But the mismatching versions itself is related to the ceph-mds process restarting without being able to flush mdlog (hitting an assert, etc..)?

From https://github.com/ceph/ceph/pull/49970, yeah this is one case could cause the mismatching.

**Comment by vshankar** [comment]
> [...]
> 
> > > @vshankar If the mds couldn't replay the journal successfully when the inodetable/sessionmap version mismatched it will skip applying the prealloc `ino#` list, which will remove the `ino#` from this list. But the inode has already been added to the `inode_map` in earlier time when relaying the journal. That means next time when we get a new `ino#` from the prealloc `ino#` list and trying to add it to the `inode_map`, it will conflict.
> > 
> > 
> > Sure. But the mismatching versions itself is related to the ceph-mds process restarting without being able to flush mdlog (hitting an assert, etc..)?
> 
> From #49970, yeah this is one case could cause the mismatching.

Right. Anyway, this change is in my list (review/test) for this week.

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
Included this is my test branch. Test runs in ~3hrs.

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230503.142424

**Comment by lxbsz** [comment]
> https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230503.142424

https://pulpito.ceph.com/vshankar-2023-05-03_16:19:11-fs-wip-vshankar-testing-20230503.142424-testing-default-smithi/7261330/

There is one failure for the test case, I will dig it.

**Comment by lxbsz** [comment]
@vshankar Just appended a new commit to fix it to make it easier to review. Please check. Thanks.

**Comment by vshankar** [comment]
> @vshankar Just appended a new commit to fix it to make it easier to review. Please check. Thanks.

Thanks, will have a look. The last run had weird pkg install issues, so, not all tests were run.

**Comment by lxbsz** [comment]
> > @vshankar Just appended a new commit to fix it to make it easier to review. Please check. Thanks.
> 
> Thanks, will have a look. The last run had weird pkg install issues, so, not all tests were run.

I have checked the failure one by one. And there are 31 failures were due to fail to install rpm packages. And found one new issue [1], and have raise one PR to fix it.

And there have two failure were caused by the issue I fixed in https://github.com/ceph/ceph/pull/43184/commits/d851a9475c30d0b6fa4418c38d04fd602bc785eb.

And all the others are known issues, such as for the quota issue and xfstest dev test failure issue, etc.

[1] https://tracker.ceph.com/issues/59657

**Comment by vshankar** [comment]
LGTM. Test runs are here: https://pulpito.ceph.com/vshankar-2023-05-06_17:28:05-fs-wip-vshankar-testing-20230506.143554-testing-default-smithi/

**Comment by vshankar** [comment]
> LGTM. Test runs are here: https://pulpito.ceph.com/vshankar-2023-05-06_17:28:05-fs-wip-vshankar-testing-20230506.143554-testing-default-smithi/

Mostly looks good. There are some unrelated failures for which I've created trackers. Will merge this soon.

**Comment by vshankar** [comment]
jenkins test make check arm64

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#09-May-2023

**Comment by vshankar** [comment]
jenkins test make check arm64

**Pull request merged**